Add Diagnostic Messages Infrastructure

natidea opened the pull request:
Add plumbing to get diagnostic messages from SDK tasks to console out put, VS Error List, and VS dependency node in solution explorer.

Part of #26

### Summary
- This is similar to the [diagnostic message APIs](https://github.com/dotnet/cli/blob/6fcbefa4f7a0016a68d3cda52779298a5cd20837/src/Microsoft.DotNet.ProjectJsonMigration/Microsoft.DotNet.Internal.ProjectModel/DiagnosticMessage.cs) in the CLI
- Uses MSBuild to log warnings and errors
- Creates `<DiagnosticMessages>` items that can be processed downstream. Right now the only downstream task is `PreprocessPackageDependenciesDesignTime` which I’ve updated to display relevant diagnostics in the dependency node.

![image](https://cloud.githubusercontent.com/assets/7732033/21242090/7efbd0ce-c2c7-11e6-8802-86596c4db3f9.png)

Here is an overview of the design:

1. I added a `DiagnosticMessages` property to the `TaskBase` class. 

```csharp
        [Output]
        public ITaskItem[] DiagnosticMessages
        {
            get { return _diagnostics.GetDiagnosticMessages(); }
        }
```

This means any task in our SDK can emit `<DiagnosticMessages>`. I use this approach because the SDK distributes functionality over many small tasks, so relevant diagnostics may come from many places. 

2. The base class also has a `DiagnosticsHelper` that contains overloaded Add methods. (Alternatively, we could add another base class to the hierarchy.) 

A task can emit a diagnostic by calling `Diagnostics.Add(...);`
This does two things:
- Add the diagnostic to a collection of `<DiagnosticMessages>`.
- If severity is warning or error, log it to MSBuild. We are logging here for convenience. Alternatively, we could let some downstream task aggregate all the diagnostics and log them.

Targets that call the task should add DiagnosticMessages to the output list:
```xml
	<Output TaskParameter="DiagnosticMessages" ItemName="DiagnosticMessages" />
```

And add themselves to a property that indicates they emit diagnostics so downstream callers can DependsOnTargets them:
```xml
  <PropertyGroup>
    <EmitsDiagnosticMessages>
      RunResolvePackageDependencies
    </EmitsDiagnosticMessages>
  </PropertyGroup>
```

3. Downstream tasks DependsOnTargets `EmitsDiagnosticMessages`. E.g.
```xml    
<Target Name="ResolvePackageDependenciesDesignTime" DependsOnTargets="$(EmitsDiagnosticMessages)">
```

Right now the only target that does this is `ResolvePackageDependenciesDesignTime`. It only considers diagnostic messages that have a `ParentTarget` and `ParentPackage` (i.e. diagnostics that appear in the dependencies node hierarchy). I have not made any changes to project system to recognize diagnostic messages, so they appear as Unknown

Still Pending:
- [x] Support messages on the TargetFramework Node
- [x] Update tests, particularly for `PreprocessPackageDependenciesDesignTime`, which currently fail the build

natidea committed:
Add plumbing to get diagnostic messages from SDK tasks to console output, VS Error List, and VS dependency node in solution explorer.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.PackageDependencyResolution.targets:
```diff
@@ -73,6 +73,14 @@ Copyright (c) .NET Foundation. All rights reserved.
     </PrepareResourcesDependsOn>
   </PropertyGroup>
 
+  <!-- List of targets that emit DiagnosticMessages items -->
```
Are all of the diagnostic messages intended to be around dependency resolution. It feels weird to have a list of targets that can warn be a requirement of ResolvePackageDependenciesDesignTime. Just about every target has the possibility of logging an error or warning that can go to the error list, but most should not run every time ResolvePackageDependenciesDesignTime.

If the pattern we want is for dependency-related diagnostics to use this and straight Log.LogError|Warning for the rest, then we should make the names for these Diagnostics less general: DependencyDiagnostic* or something.

natidea replied:
I agree that as I was implementing this, it became clear that currently only diagnostics that surface in the dependency node really need this infrastructure. Other diagnostics can just Log.LogError|Warning.

On the other hand, a collection of named Diagnostic items that can be manipulated downstream could be useful for as yet unimagined purposes. I'm also wondering if there is any distinction between the coded diagnostics listed in [NuGet](https://github.com/dotnet/cli/blob/d7b0190bd4ce49121a7cd5bdc0423ab85229a085/src/Microsoft.DotNet.ProjectModel/ErrorCodes.NuGet.cs)/[CLI](https://github.com/dotnet/cli/blob/d7b0190bd4ce49121a7cd5bdc0423ab85229a085/src/Microsoft.DotNet.ProjectModel/ErrorCodes.DotNet.cs), and errors/warnings we already emit in our tasks. The former seems dedicated to helping users diagnose why things may not be working as expected, while the latter are more general failures such as a null reference or some other unexpected value. 

I'm open to implementing it either way. Perhaps a compromise is to have an additional property `EmitsDependencyDiagnostic` that is limited to the subset of targets that emit relevant diagnostics.

natidea committed:
Make InputDiagnosticMessages optional

natidea committed:
Merge tracking branch upstream/master into diagnostics

natidea committed:
Remove checks for parent target/package
Null parent targets are already ignored when building dependency tree; null parent packages signal the warning should appear at the top level of the tree

natidea committed:
Add TFM to diagnostic id, so diagnostics can be differentiated by TFM. Also fixes error in computation of id.

natidea committed:
Add a pair of diagnostics to detect unresolved packages and unexpected dependency versions

natidea committed:
Suppress diagnostics from error list that are already reported by NuGet (but still report them in solution explorer). Long term, NuGet will add to these to lock file and they will simply be raised here (https://github.com/NuGet/Home/issues/1599)

natidea committed:
Rename EmitsDiagnosticMessages to EmitsDependencyDiagnosticMessages

natidea commented:
/cc @nguerrera @dsplaisted  could you take another look?

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/PreprocessPackageDependenciesDesignTime.cs:
```diff
@@ -63,6 +68,9 @@ protected override void ExecuteCore()
 
             PopulateAssemblies();
 
+            InputDiagnosticMessages = InputDiagnosticMessages ?? new ITaskItem[] { };
```
nit: `Array.Empty<ITaskItem>()`

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/PreprocessPackageDependenciesDesignTime.cs:
```diff
@@ -52,6 +54,9 @@ public class PreprocessPackageDependenciesDesignTime : TaskBase
 
+        private Dictionary<string, ItemMetadata> DiagnosticsMap { get; set; }
+                    = new Dictionary<string, ItemMetadata>(StringComparer.OrdinalIgnoreCase);
+
```
Pre-existing, but why do we have private properties instead of private fields with `_name`?

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/PreprocessPackageDependenciesDesignTime.cs:
```diff
@@ -166,6 +176,16 @@ private void PopulateAssemblies()
+        {
+            foreach (var diagnostic in InputDiagnosticMessages)
+            {
+                // TODO generate unique ID for each diagnostic as target/package/code may not be unique enough
```
File bug for any TODO

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/PreprocessPackageDependenciesDesignTime.cs:
```diff
@@ -334,5 +354,48 @@ private class AssemblyMetadata : PackageMetadata
             }
         }
+
+        private class DiagnosticMetadata : ItemMetadata
```
nit: sealed

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/DiagnosticsHelper.cs:
```diff
@@ -0,0 +1,153 @@
+    /// Writes diagnostic messages to the task log and creates diagnostic task items 
+    /// that can be returned from a task
+    /// </summary>
+    public class DiagnosticsHelper
```
sealed

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/DiagnosticsHelper.cs:
```diff
@@ -0,0 +1,153 @@
+                        columnNumber    : startColumn,
+                        endLineNumber   : endLine,
+                        endColumnNumber : endColumn, 
+                        message         : message);
```
Nit: Does VS preserve this formatting on Ctrl+K, Ctrl + F? 

I'd prefer we did not get cute lining things up against the grain. It causes unnecessary diffs when something longer than all of the others gets added.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/LockFileExtensions.cs:
```diff
@@ -54,6 +54,29 @@ public static LockFileTargetLibrary GetLibrary(this LockFileTarget lockFileTarge
+
+        public static Dictionary<string, string> GetProjectFileDependencies(this LockFile lockFile)
+        {
+            Dictionary<string, string> projectDeps = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
```
nit: var

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageDependencies.cs:
```diff
@@ -399,6 +403,56 @@ private void GetFileDependencies(LockFileTargetLibrary package, string targetNam
+                        1, 0,
+                        target.Name,
+                        packageId,
+                        logToMSBuild: false // Temporarily suppress MSBuild logging because this diagnostic is also reported by NuGet
```
Is there a bug tracking this on nuget side + one on our side to follow up when that's done?

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.PackageDependencyResolution.targets:
```diff
@@ -69,6 +69,14 @@ Copyright (c) .NET Foundation. All rights reserved.
+      RunResolvePackageDependencies;
+      $(EmitsDependencyDiagnosticMessages)
+    </EmitsDependencyDiagnosticMessages>
+  </PropertyGroup>
```
Generally the property for depends on is XxxDependsOn, but we've also moved away from that in favor of BeforeTargets. I wonder if we shouldn't just have things use BeforeTargets=ResolvePackageDependencies and skip having this list? What else do we think will go here?

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.PackageDependencyResolution.targets:
```diff
@@ -170,14 +179,15 @@ Copyright (c) .NET Foundation. All rights reserved.
   <Target Name="ResolvePackageDependenciesDesignTime"
           Returns="@(_DependenciesDesignTime)"
-          DependsOnTargets="ResolveAssemblyReferencesDesignTime;RunResolvePackageDependencies">
+          DependsOnTargets="ResolveAssemblyReferencesDesignTime;RunResolvePackageDependencies;$(EmitsDependencyDiagnosticMessages)">
```
Why is RunResolvePackageDependencies hard-coded here and also in EmitsDependencyDiagnosticMessages?

nguerrera reviewed (approved):
Looks good. Minor feedback only.

dsplaisted reviewed (approved):
There is a TODO at the top of PreprocessPackageDependenciesDesignTime.cs that you can remove now.

Also, it was hard for me to follow what PreprocessPackageDependenciesDesignTime is doing.  It would probably help to have comments explaining how the items it outputs are structured in order to build the tree.

natidea committed:
Add comment links to NuGet/SDK bugs for getting diagnostics from lockfile, and other PR feedback

natidea commented:
Updated with PR feedback and filed bug to remove MSBuild logging suppression
/cc @srivatsn for approval

natidea committed:
Correct message when Project reference is unresolved
